### PR TITLE
Simplify the  `PushDownLimit`.

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?

None. This is a follow-up of #4411 and  #4276.

# Rationale for this change
Rewrite the code to:
1. Add more comments for `merge limits`
2. Return `None` in `try_optmize` if the plan can't be optimized.
3. Simplify the code.

# What changes are included in this PR?

# Are these changes tested?
No functional changes, so don't need new tests.

# Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by alamb** [comment]
I plan to review this PR tomorrow -- I apologize for the delay @HaoYang670

File: datafusion/optimizer/src/push_down_limit.rs [hunk 1]
```diff
@@ -37,104 +37,97 @@ impl PushDownLimit {
...
         let limit = match plan {
             LogicalPlan::Limit(limit) => limit,
             _ => return Ok(None),
         };
 
-        if let LogicalPlan::Limit(child_limit) = &*limit.input {
+        if let LogicalPlan::Limit(child) = &*limit.input {
+            // Merge the Parent Limit and the Child Limit.
+
+            //  Case 0: Parent and Child are disjoint. (child_fetch <= skip)
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_limit.rs:57]: these pictures are wonderfully helpful -- thank you @HaoYang670

**Review by alamb — APPROVED** [review]
Thanks @HaoYang670  -- I went through this logic carefully and it makes sense to me. Thank you for the wonderful diagram as well. 👍 

cc @jackwener @ming535 and @liukun4515

File: datafusion/optimizer/src/push_down_limit.rs [hunk 2]
```diff
@@ -238,11 +218,57 @@ impl OptimizerRule for PushDownLimit {
...
     } else {
-        fetch - skip
+        match join.join_type {
+            Left => (Some(limit), None),
+            Right => (None, Some(limit)),
+            _ => (None, None),
+        }
+    };
+
+    if let (None, None) = (left_limit, right_limit) {
```
Comment (Dandandan) [inline datafusion/optimizer/src/push_down_limit.rs:243]: I think this would be more clear with `match` here than using `if let`

**Review by jackwener — APPROVED** [review]
LGTM, Thanks for @HaoYang670

File: datafusion/optimizer/src/push_down_limit.rs (hunk 2 above)
Comment (HaoYang670) [reply datafusion/optimizer/src/push_down_limit.rs]: Updated.

**Comment by alamb** [comment]
Thanks again all!

**Pull request merged**